auth/oauth.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022 MongoDB Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"runtime"
+	"strings"
+
+	atlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+const defaultBaseURL = atlas.CloudURL
+
+var (
+	userAgent = fmt.Sprintf("go-mongodbatlas/%s (%s;%s)", atlas.Version, runtime.GOOS, runtime.GOARCH)
```

Comment:
[possible future improvement] use the same user agent across the codebase
https://github.com/mongodb/go-client-mongodb-atlas/blob/80a3a1b7eb84c9224e648a1ea30ff9d1ffd44850/mongodbatlas/mongodbatlas.go#L48